include/ofi_mr.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2017 Intel Corporation, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+
+#include "fi.h"
+#include "fi_util.h"
+
+/*
+ * MR cache
+ */
+
+struct ofi_mr_cache_attr {
+	int			reg_size;
+	int			stale_size;
+	int			elem_size;
+	struct fi_provider	*prov;
+	/*struct ofi_mr_monitor	*mr_monitor;*/
+	void			*reg_context;
+	void			*dereg_context;
+	void *(*reg_callback)(void *handle, void *address, size_t length,
+			      void *reg_context, void *context);
+	int (*dereg_callback)(void *handle, void *context);
+};
```

Comment:
- Remove structure and move the necessary fields directly into of_mr_cache
- The reg_context and dereg_context fields/parameters shouldn't be necessary
- Replace the provider field with a pointer to a domain

Reply:
> Remove structure and move the necessary fields directly into of_mr_cache

the fields of the `ofi_mr_cache_attr` structure have been move into the `ofi_mr_cache`. but how do we pass the args and callback into the init function? Directly?

>The reg_context and dereg_context fields/parameters shouldn't be necessary
Removed, but we need to pass domain to the reg/dereg_callbacks. it's needed for the provider's function that is used to register MR

>Replace the provider field with a pointer to a domain
Done

Reply:
I was just going to have the caller initialize the ofi_mr_cache fields directly.  There doesn't seem to be any reason for the caller to set attr.some_field, then call a function that copies the attribute structure, versus just having the caller set cache.field instead.

I would have the callbacks take the cache structure as input.

Reply:
Okay, I got you idea.

@shefty Should we sanitize the user's input? I.e. to check that all callbacks are provided?

Reply:
Yes, I would add asserts to some init() call that these have been set.

Reply:
> Yes, I would add asserts to some init() call that these have been set.

Done. I've updated the PR to apply this set of the comments